### PR TITLE
:tada: Add typography token row

### LINKED ISSUE #9932: Add composite typography token input to the Design sidebar
https://tree.taiga.io/project/penpot/us/11964

Release: 2.17

Flag: `enable-token-typography-row`

As a designer, I want to apply **composite typography tokens** directly from the Design sidebar so that text styles stay consistent, reusable, and easy to update across files and teams.

This US can be activated using `enable-token-typography-row` flag

**Context:**

Today text properties (font family, size, weight, etc.) are edited as raw values; there is no direct way to bind them to typography tokens. This story covers **composite typography tokens only** — single/atomic tokens will be addressed in a separate user story.

A composite typography token bundles multiple text properties at once (font-family, font-size, font-weight, line-height, letter-spacing, text-case, text-decoration).

There are existing explorations for the composite input in this Penpot file.

#### Interaction

This interaction only applies when **composite typography tokens** exists in the file.

**Accessing the token picker**

A new button appears in the **Design sidebar → TEXT section**, next to the "+" button used to create text styles. This button is only visible when composite typography tokens exist in the file.

Clicking the button opens a **dropdown token picker** with:

*   A **search input** to filter tokens by name.
*   A title: **"Typography Tokens"** (ES: _Tokens Tipográficos_), to make clear that only composite tokens are listed here — not single/atomic ones.
*   A **flat list** of all available composite typography tokens (no grouping for now — TDD).
*   Each item shows the **token name**; on hover, a **tooltip** displays the full token name and its resolved values.

**Applying a token**

When the user clicks a token from the list:

*   The dropdown closes.
*   The token is applied to the selected text layer.
*   The TEXT block in the Design sidebar updates to show the **token input state** (see below).

**Token input (applied state)**

When a composite token is applied, the new token input **replaces** the following controls:

*   Font family input
*   Font size select
*   Font weight/style select
*   Line height input
*   Letter spacing input
*   Text case buttons

Typography tokens also define text decoration (underline/strikethrough), so in expanded options, the **font style buttons appear disabled** and display the currently applied value.

Behavior of the token input:

*   **On hover:** shows an icon to **detach** the token.
*   **Detaching** the token preserves the current applied values — it does not reset them.
*   **Clicking on individual text property controls** does not detach the token, since those properties are driven by the token.
*   **To change the applied token**, the user uses the same token button next to the "+" icon.
*   **On hover**, a tooltip shows the token name and all resolved properties.

**Resolved value tooltip**

On hover over the token input (or token items in the list):

*   Token type: "Composite typography token"
*   Token name
*   Key resolved values: font-family, font-size, line-height, letter-spacing, text-case, text-decoration

Typography tokens also defines the text style (underline/striketrough) so, in expanded options, the font style buttons will appear disabled (and display the applied value if there is any:

*   On hover, the input will display an icon to dettach the token

*   Detaching the token won't change the current values applied to the
*   Clicking on any of the text properties won't dettach the token as they are not part of the typography token.
*   To change the applied token, the user should use the token button next to + icon (for now)
*   On hover, display a tooltip with the token name and the resolved properties.

**Invalid states / not active sets**

While using tokens, there are some cases that can happen while a token is applied.

*   There is no active set / theme so that the applied token is not active - [Display a unresolved token state](https://design.penpot.app/#/workspace?team-id=52961d58-0a92-80c2-8003-38569623b41c&file-id=502c5b43-61ea-81d3-8004-88ad2e9a68fa&page-id=e65eee7a-479c-8076-8006-bc420484f5d1&layout=layers&board-id=5fda8e42-db3e-8019-8006-cff4283fd14d)
*   The current token is broken / has an invalid value. This can happen in some corner cases - [Display a unresolved token state](https://design.penpot.app/#/workspace?team-id=52961d58-0a92-80c2-8003-38569623b41c&file-id=502c5b43-61ea-81d3-8004-88ad2e9a68fa&page-id=e65eee7a-479c-8076-8006-bc420484f5d1&layout=layers&board-id=5fda8e42-db3e-8019-8006-cff4283fd14d) for now.
*   if the composite token has been removed - [Display a unresolved token state](https://design.penpot.app/#/workspace?team-id=52961d58-0a92-80c2-8003-38569623b41c&file-id=502c5b43-61ea-81d3-8004-88ad2e9a68fa&page-id=e65eee7a-479c-8076-8006-bc420484f5d1&layout=layers&board-id=5fda8e42-db3e-8019-8006-cff4283fd14d) for now - with a specific tooltip: `This token does not exists or has been deleted.` (Este token no existe o ha sido eliminado). If this token is created again or imported, it will start working again.

* * *

Individual inputs are defined [in this user story](https://tree.taiga.io/project/penpot/us/13569)

* * *

### 🤔 Unknowns and challenges

**New UI for single typography tokens (Group B)**          
  - We need to define a UI pattern for applying `font-family`, `font-size`, `font-weight`, `text-case`, and `text-decoration` tokens, which does not yet exist elsewhere in the app.          
  - This will likely require research on how similar tools (design systems, design tools, code editors) expose typography tokens or styles, to avoid creating an inconsistent or confusing control.          
  - The outcome of this research may influence both the visual design and the interaction model (search, filtering, previews, etc.) for these inputs.

**Resolved value tooltip design**          
  - Exact formatting for token type, property names, and values.          
  - Whether to mirror the `Resolved value: [#ff0000](project/penpot/t/ff0000)` pattern or use richer multi-line typography detail.          
  - How to handle long composites without overwhelming the tooltip.

**Overrides vs binding behavior**          
  - Final rule when editing a token-driven property (unbind vs marked override) and visual indication.

**Token scope and conflicts**          
  - How to represent scopes (file/project/library) in pickers.          
  - Behavior when a bound token is renamed or deleted.

* * *

### ☕ Requirements

\[x\] The User Guide will need update (new typography token controls and tooltips)          
\[x\] New translations needed (labels/tooltips for new inputs and states)

* * *

### 🖥️ How to test
**No typography tokens in the file**
- Create a new file.
- Check that there is no tokens button next to the "Text" section title in the design tab.

**Existing tokens in the file**
- Add a typography token.
- Check that a button now appears next to the "Text" section title.

**Application**
- Click the tokens button.
- A search bar is shown.
- A list of sets with tokens is shown.

**Apply composite token**
- Apply a composite token; verify all related properties and tooltip values match the token.

**Search and apply**
- Click the tokens button.
- Search for a token by name and check that the list is filtered.
- Apply a token.
- Check that the menu is closed and the token is applied to the block element.
- The + button (to create a style) disappears.

**Apply single tokens over a typography token**
- Apply a typography token to a text element.
- Apply a font size token to the same element.
- Check that the composite token has been unapplied and the font size token is now applied.

**Edit text**
- Apply a token to a text layer.
- Edit the content of the text layer.
- Confirm that the token has not been unapplied.

**Property edits**
- Edit a property after applying a token (text alignment, box behavior, etc.); confirm the agreed override behavior and UI feedback.
- Confirm that the token has not been unapplied.

**Font style styles**
- Apply a typography token to a text
- Check that text styles are disabled and cannot be changed (underline, strikethough - since those are part of the typography token)

**Apply a typography style (asset)**
- Apply a typography token to a text
- Apply a typography sytle (assets)
- check that token has been unapplied.

**Tooltip**
- Apply a typography token to a text
- Confirm that the tooltip information corresponds to the token properties
- edit token properties
- check that the tooltip info in the input has the udated information.

**Error and edge cases**
- Test missing/renamed tokens, very long token names, and large composites.
- Verify keyboard navigation and screen-reader output for inputs and tooltips.

---

Main fixes: typos (*Cllick*, *dissapears*, *typògraphy*, *elemente*), inconsistent bullet styles (mixed `*` and `-`), inconsistent capitalisation and punctuation, and a few minor phrasing tweaks for clarity.
* * *

*   User story for [TYPOGRAPHY COMPOSITE INPUT](https://tree.taiga.io/project/penpot/us/11964)
*   User story for [INDEPENDENT TYPOGRAPHY INPUTS](https://tree.taiga.io/project/penpot/us/13569) ⚪

Design: https://design.penpot.app/#/workspace?team-id=52961d58-0a92-80c2-8003-38569623b41c&file-id=502c5b43-61ea-81d3-8004-88ad2e9a68fa&page-id=e65eee7a-479c-8076-8006-bc420484f5d1 
External source: 
External source 2:

### PR DESCRIPTION
**Pull request by EvaMarco**
### Related Ticket

This closes [ this US ](https://tree.taiga.io/project/penpot/us/11964?milestone=510141)
### Summary

This PR introduces a new component called typography-row, which replaces the individual typography inputs with a unified row when a typography token is applied.

### Key features include:

A button that opens a dropdown with all available typography tokens.
A built-in search input within the dropdown to filter and quickly find tokens.
The ability to select and apply a typography token from the dropdown.
An option to detach the currently applied token.

### Behavior
The typography-row is displayed when a typography token is applied, replacing the standard typography inputs.
Users can open the dropdown, search, and select a token seamlessly.
If a token is already applied, users can also detach it and return to manual typography controls.
#### Requirements / Limitations
This feature only works with text shapes.
A typography token must exist in order to use the dropdown.
The feature is behind the `enable-typography-token-row` feature flag.

### Steps to Reproduce
- Enable the `enable-typography-token-row` feature flag in the configuration.
- Ensure at least one typography token is created.
- Select a text shape.

#### Verify that:
- The typography-row replaces the default typography inputs when a token is applied.
- The dropdown button opens the list of typography tokens.
- The search input filters the available tokens.
- Selecting a token applies it correctly.
- The detach option removes the applied token.

### Checklist

- [ ] Choose the correct target branch; use `develop` by default.
- [ ] Provide a brief summary of the changes introduced.
- [ ] Add a detailed explanation of how to reproduce the issue and/or verify the fix, if applicable.
- [ ] Include screenshots or videos, if applicable.
- [ ] Add or modify existing integration tests in case of bugs or new features, if applicable.
- [ ] Refactor any modified SCSS files following the refactor guide.
- [ ] Check CI passes successfully.
- [ ] Update the `CHANGES.md` file, referencing the related GitHub issue, if applicable.

### CODE REVIEW

File: frontend/src/app/main/ui/workspace/sidebar/options/menus/text.cljs [hunk 1]
```diff
@@ -241,117 +343,165 @@
...
+             (st/emit! (dwu/start-undo-transaction uid))
+             (when (features/active-feature? @st/state "text-editor/v2")
+               (let [content (when editor-instance
+                               (content/dom->cljs (dwt/get-editor-root editor-instance)))]
+                 (when (some? content)
+                   (st/emit! (dwt/v2-update-text-shape-content (first ids) content :finalize? true)))))
+             (st/emit! (dwsh/update-shapes ids #(assoc % :grow-type grow-type)))
+             (when (features/active-feature? @st/state "render-wasm/v1")
+               (st/emit! (dwwt/resize-wasm-text-all ids)))
+             (ts/schedule #(st/emit! (dwu/commit-undo-transaction uid))))))
```
Comment (hirunatan) [inline frontend/src/app/main/ui/workspace/sidebar/options/menus/text.cljs:374]: Is this really from this PR? Seems quite unrelated. Could it be a reminder of a rebase with conflicts?

File: frontend/src/app/main/ui/ds/controls/shared/searchable_options_dropdown.cljs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        (when-let [list (mf/ref-val list-ref)]
+          (when-not (dom/active? list)
+            (dom/focus! list)))
+        (when-let [node (obj/get (mf/ref-val nodes-ref) focused-id)]
+          (dom/scroll-into-view-if-needed! node {:block "nearest" :inline "nearest"}))))
+
+    [:> :ul list-props
+     [:li {:class (stl/css :option-search)
+           :role  "presentation"}
+      ;; TODO; review this placeholder.
```
Comment (hirunatan) [inline frontend/src/app/main/ui/ds/controls/shared/searchable_options_dropdown.cljs]: TODO?

File: frontend/src/app/main/ui/workspace/sidebar/options/menus/text.cljs (hunk 1 above)
Comment (EvaMarco) [reply frontend/src/app/main/ui/workspace/sidebar/options/menus/text.cljs:374]: This was moved from the grow component, it was called handle-change-grow. 
I try to make all subcomponents work the same way.

Comment (EvaMarco) [reply frontend/src/app/main/ui/workspace/sidebar/options/menus/text.cljs:374]: It was on line 134 in the red side

Comment (hirunatan) [reply frontend/src/app/main/ui/workspace/sidebar/options/menus/text.cljs:374]: Ahh perfect. Ok then.

**Review by hirunatan — APPROVED** [review]
